components/script/script_thread.rs

```diff
@@ -1997,45 +1999,67 @@ impl ScriptThread {
         window.evaluate_media_queries_and_report_changes();
     }
 
-    /// Initiate a non-blocking fetch for a specified resource. Stores the InProgressLoad
+    /// Instructs the constellation to fetch the document that will be loaded. Stores the InProgressLoad
     /// argument until a notification is received that the fetch is complete.
-    fn start_page_load(&self, incomplete: InProgressLoad, mut load_data: LoadData) {
+    fn pre_page_load(&self, incomplete: InProgressLoad, load_data: LoadData) {
         let id = incomplete.pipeline_id.clone();
-
-        let context = Arc::new(Mutex::new(ParserContext::new(id, load_data.url.clone())));
-        let (action_sender, action_receiver) = ipc::channel().unwrap();
-        let listener = NetworkListener {
-            context: context,
-            task_source: self.networking_task_source.clone(),
-            wrapper: None,
-        };
-        ROUTER.add_route(action_receiver.to_opaque(), box move |message| {
-            listener.notify_fetch(message.to().unwrap());
-        });
-
-        if load_data.url.scheme() == "javascript" {
-            load_data.url = ServoUrl::parse("about:blank").unwrap();
-        }
-
-        let request = RequestInit {
+        let mut req_init = RequestInit {
             url: load_data.url.clone(),
             method: load_data.method,
             destination: Destination::Document,
             credentials_mode: CredentialsMode::Include,
             use_url_credentials: true,
-            origin: load_data.url,
+            origin: load_data.url.clone(),
             pipeline_id: Some(id),
-            referrer_url: load_data.referrer_url,
+            referrer_url: load_data.referrer_url.clone(),
             referrer_policy: load_data.referrer_policy,
-            headers: load_data.headers,
+            headers: load_data.headers.clone(),
             body: load_data.data,
+            redirect_mode: RedirectMode::Manual,
             .. RequestInit::default()
         };
 
-        self.resource_threads.send(CoreResourceMsg::Fetch(request, action_sender)).unwrap();
+        if req_init.url.scheme() == "javascript" {
+            req_init.url = ServoUrl::parse("about:blank").unwrap();
+        }
+
+        let mut res_init = ResponseInit {
+            url: load_data.url.clone(),
+            headers: Headers::new(),
+            referrer: load_data.referrer_url,
+        };
+        res_init.headers.set::<Location>(Location(load_data.url.to_string()));
+
+        self.constellation_chan.send(ConstellationMsg::InitiateNavigateRequest(req_init, res_init, id)).unwrap();
         self.incomplete_loads.borrow_mut().push(incomplete);
     }
 
+    /// Begin parsing.
+    fn start_page_load(&self,
```

Comment:
We will want to split this into three different handlers. The first one will need to create the ParserContext and store it somewhere accessible (like a vector in ScriptThread), and the others will retrieve it and update it (ie. process_response_chunk and process_response_eof). Perhaps inside `incomplete_loads`?